### DISCUSSION TITLE
Wav2vec with LM

### DISCUSSION THREAD

**Discussion by 73minerva** [speechbrain/speechbrain#1661] [Q&A]
Hi,
I have read the wav2vec recipe for CV dataset and I was wondering if there are any implementation and experiments that had wav2vec alongside a Transformer LM ?! Does it work as a proper implementation If I change the decoder in the config file?!

**Comment by TParcollet** [comment]
Alright, we never tried but it is feasible. You would have to 1. fine-tune a wav2vec 2.0 with BPE; 2. Train a transformer LM with the same BPE; 3. Combine both in the decoding. You can also use a n-gram model directly with ctc-decode or py-ctc-decode. Important note: We are on the refactoring of this part. By the end of the summer, we will have a brand new integration of LM fusion and rescoring that will ease everything.

**Comment by Craya** [comment]
Hi @TParcollet ,

We have the exact same need as @73minerva, do you still confirm the target date of end of summer for LM integration?

We don't want to spend hours this summer on a ineficient internal LM that would be overwhelmed by your future release...

Thanks for your feedback

Fabien.

**Comment by TParcollet** [comment]
We added many things to the next major version, so it will most likely be postponed to Fall. Note that a pyctcdecode integration would be extremely simple to do (literally just to call the function on our CTC logits) if you wish to add an n-gram model. If it's RNNLM or Transformer LM and CTC only, well, you may need to wait indeed. One solution here is to follow the "Unstable" branch, so you can be pinged as soon as we refactor the rescoring part. Basically, the code is ready, but the review and integration process of such a large part takes time.

**Comment by Craya** [comment]
Thanks @TParcollet ,

We already have a n-gram model, so we will follow your advice and have a look to pyctcdecode.
I can't wait for Fall to play with the next version!

A bientôt.

Fabien.

**Comment by VitoPepe** [comment]
Hi @TParcollet 

I tried your method with pyctcdecode to integrate a ngram model. However when I decode my file with the decode function of pyctcdecode, the transcription that I get is incoherent. Do you have an idea of why my pyctcdecode integration is bad ? 

```
ngram_lm_path = "eng_lm.arpa"

print("## Load fine-tuned ENG ASR:")
asr_eng_model = EncoderASR.from_hparams(source=model_path, savedir=model_path,hparams_file="hyperparams_inference.yaml" ,run_opts={"device":"cuda"})

def get_decoder_ngram_model(tokenizer, ngram_lm_path, vocab_path=None):
    vocab_list = [[tokenizer.id_to_piece(id), id] for id in range(tokenizer.get_piece_size())]
    vocab_dict = {}
    for rec in vocab_list:
        k = rec[0] #dict key
        v = rec[1:] #dict value, convert to list if needed
        vocab_dict[k] = v # set key and value
    
    sort_vocab = sorted((value, key) for (key, value) in vocab_dict.items())
    vocab = [x[1] for x in sort_vocab]
    vocab_list = vocab

    decoder = build_ctcdecoder(vocab_list, ngram_lm_path)
    return decoder

ngram_lm_model = get_decoder_ngram_model(asr_eng_model.tokenizer, ngram_lm_path)

def transcribe_file(path, max_seconds=120):
    waveform = asr_eng_model.load_audio(path)
    batch = waveform.unsqueeze(0)
    rel_length = torch.tensor([1.0])
    with torch.no_grad():
        logits = asr_eng_model(batch, rel_length)
    text_batch = [ngram_lm_model.decode(
        logit.detach().cpu().numpy()) for logit in logits]
    return text_batch[0]

filename = "audio.wav"
print("lm: " + transcribe_file(filename))
```

Thanks for you feedback, 

Simon

**Comment by thibault-roux** [comment]
I'm really interested in rescoring with a RNNLM! Can you tell me where I can find the yaml for wav2vec + LM in the "Unstable"? I didn't find it in the Common Voice folder.

**Reply by VitoPepe** [reply]
You're looking for the .yml use a LM in your training ? I've already made an implementation of a language model for the inferences but would be interested in implementing it for training as well

**Reply by thibault-roux** [reply]
I was actually looking for the implementation of LM for inference! Is it possible that you send it to me? You can send me an e-mail at thibault.roux@univ-nantes.fr if it is convenient to you.

**Reply by anautsch** [reply]
The rescoring is WIP #751 - it's not yet merged anywhere.

@Adel-Moumen maybe you know more on this.
